src/main/kotlin/snyk/oss/annotator/OSSBaseAnnotator.kt

```diff
@@ -0,0 +1,131 @@
+
+                val severity = severity(vulnerability)
+                val textRange = textRange(psiFile, vulnerability)
+                val annotationBuilder =
+                    holder.newAnnotation(severity, annotationMessage(vulnerability)).range(textRange)
+                val fixRange = fixRange(psiFile, vulnerability)
+                if (fixRange != TextRange.EMPTY_RANGE && remediation != null && remediation.upgrade.isNotEmpty()) {
+                    addQuickFix(psiFile, vulnerability, annotationBuilder, fixRange, remediation)
+                }
+                annotationBuilder.create()
+            }
+    }
+
+    open fun getIssuesForFile(psiFile: PsiFile): OssVulnerabilitiesForFile? {
+        val ossResult = getSnykToolWindowPanel(psiFile.project)?.currentOssResults
+        val fileName = psiFile.virtualFile.presentableUrl
+
+        ProgressManager.checkCanceled()
+
+        return ossResult?.allCliIssues
+            ?.filter { AnnotatorHelper.isFileSupported(fileName) }
+            ?.firstOrNull { fileName.endsWith(it.displayTargetFile.replace("-lock","")) }
+    }
+
+    open fun annotationMessage(vulnerability: Vulnerability): String {
+        return buildString {
+            append("Snyk: ${vulnerability.title} in ${vulnerability.name}")
+        }
+    }
+
+    open fun addQuickFix(
+        psiFile: PsiFile,
+        vulnerability: Vulnerability,
+        annotationBuilder: AnnotationBuilder,
+        textRange: TextRange,
+        remediation: OssVulnerabilitiesForFile.Remediation
+    ) {
+        val replacementText = getFixVersion(remediation, vulnerability)
+        if (replacementText.isNotBlank()) {
+            annotationBuilder.withFix(AlwaysAvailableReplacementIntentionAction(textRange, replacementText))
+        }
+    }
+
+    open fun getFixVersion(remediation: OssVulnerabilitiesForFile.Remediation, vulnerability: Vulnerability): String {
+        val upgrade = remediation.upgrade[vulnerability.from[1]]
+        return upgrade?.upgradeTo?.split("@")?.get(1) ?: ""
+    }
+
+    open fun textRange(psiFile: PsiFile, vulnerability: Vulnerability): TextRange {
+        val document = psiFile.viewProvider.document ?: return TextRange.EMPTY_RANGE
+        val lines = document.text.lines()
+        var lineStart = 0
+        var lineEnd = 0
+        var colStart = 0
+        var colEnd = 0
+        for (i in lines.indices) {
+            val line = lines[i]
+            if (lineMatches(psiFile, line, vulnerability)) {
+                lineStart = i
+                lineEnd = i
+                colStart = colStart(line, vulnerability)
+                colEnd = colEnd(line, vulnerability)
+                break
+            }
+        }
+        val lineOffSet = document.getLineStartOffset(lineStart) + colStart
+        val lineOffSetEnd = document.getLineStartOffset(lineEnd) + colEnd
+        return TextRange.create(lineOffSet, lineOffSetEnd)
+    }
```

Comment:
all that text parsing above and below seems to be too error-prone... What abot text in comments? in other sting? accidental/partial matches?...

Reply:
This is a base class - by overwriting the defaults for `lineMatches()`, `colstart()` and `colend()`you can actually use a language parser to do this. I doubt though, that it's worth the effort to write language parsers for all package manager formats. But I totally agree, it's never a 100% safe unless the scanning engine itselfs reports on the place where it found the vulnerability. Else it's just guess-work.